feat: add Nlog 沒有成功產生log.txt

cycumonk opened the pull request:
feat: add Nlog 沒有成功產生log.txt

cycumonk-coder committed:
feat: add Nlog 沒有成功產生log.txt

miterfrants commented:
去看官方文件
先知道 NLog 是幹嘛用的

cycumonk-coder committed:
feat: 使用nlog紀錄打api/v1/test

cycumonk commented:
麻煩你幫我看這樣對不對

miterfrants commented on api/Program.cs:
```diff
@@ -1,37 +1,109 @@
 builder.Services.AddCors(options =>
 {
-    options.AddPolicy("AllowAllOrigins", builder =>
+    options.AddPolicy("AllowAllOrigins", policyBuilder =>
```
為什麼要改這個？

cycumonk replied:
本來搞亂跟26行的builder 但因為作用域不同，所以不改應該也沒關係

miterfrants commented on api/Program.cs:
```diff
@@ -1,37 +1,109 @@
 app.Run();
+
+
+try
```
這一段是做什麼的？

cycumonk replied:
這一段是從github範例複製下來的，整段刪掉也不影響log

miterfrants commented on api/appsettings.json:
```diff
@@ -1,9 +1,48 @@
-      "Microsoft.AspNetCore": "Warning"
-    }
-  },
+        "NLog": {
```
請解釋這個修改

cycumonk replied:
2-8 行這段是設置Nlog為紀錄log的工具
10-13 配置Nlog 基本設定 包含免關閉自動更新、日誌紀錄等級設置（Trace最低 表示全部紀錄）、Nlog內部的log檔要放的位置、配置錯誤時是否要發出異常
14-38 設置log黨的目標是什麼，這裡設置主控台跟檔案，所以在consle可以看到log檔，也可以在指定的資料夾看到log檔
25-29 配置日誌紀錄規則 並寫入console跟file 最低等級為Trace 所以全部都要紀錄

miterfrants reviewed (commented):
checked

miterfrants commented on api/Program.cs:
```diff
@@ -1,37 +1,109 @@
+    LogManager.Shutdown();
+}
+
+public class Runner
```
為什麼需要這個？

cycumonk replied:
刪掉不需要 從github範例複製下來的

cycumonk replied:
修正 如果沒有正確關閉Nlog可能會導致存在緩衝區的資料沒被寫入日誌 
也會讓記憶體被Nlog吃掉 沒辦法關閉應用程式
所以LogManager.Shutdown(); 是必要的

miterfrants reviewed (commented):
checked

miterfrants commented:
你的 console 有印東西
但是你的 log 裡頭沒新增東西

<img width="1710" alt="截圖 2024-11-09 23 35 35" src="https://github.com/user-attachments/assets/8dd8f455-91d1-42fb-954d-4f03ad9744e3">

miterfrants commented:
在關掉 application 的時候
會發生錯誤

<img width="1152" alt="截圖 2024-11-09 23 43 42" src="https://github.com/user-attachments/assets/1d5e810b-96ec-41f0-9db2-7587d2d9031a">

cycumonk commented:
<img width="921" alt="截圖 2024-11-11 晚上11 52 37" src="https://github.com/user-attachments/assets/e7af9203-d6e3-440f-8f3d-3c5a84e298d6">
console 跟 log檔都有寫入

cycumonk-coder committed:
fix : 刪除不需要的程式碼，設定fiie的位置

cycumonk-coder committed:
fix : 裡面的builder跟外面的builder作用域不同不用改

cycumonk-coder committed:
fix 正確關閉Nlog

miterfrants commented on api/Program.cs:
```diff
@@ -1,37 +1,59 @@
 app.Run();
+
+//釋放Nlog使用的資源 確保應用程式正常退出 有些日誌訊息會放在記憶體緩衝區，如果沒有正確關閉Nlog會導致資料還在緩衝區來不及寫入就被關閉，導致日誌沒寫入
+LogManager.Shutdown();
```
這個這樣寫對嗎？
不是應該在 app 關掉的時候在把 LogManager.Shutdown?

cycumonk replied:
改成判斷app stoping的時候關閉

miterfrants commented on api/appsettings.json:
```diff
@@ -1,9 +1,34 @@
+            },
+            "file": {
+                "type": "File",
+                "fileName": "/Users/huangshiwen/project/react-0812/api/logs/console-api.log",
```
我沒有這個資料夾會出現錯誤
<img width="1619" alt="截圖 2024-11-14 00 23 56" src="https://github.com/user-attachments/assets/a4242049-91d7-4444-9f7d-8de64ab16e67">

cycumonk replied:
改成在根目錄新增一個logs資料夾

miterfrants commented on api/appsettings.json:
```diff
@@ -1,9 +1,34 @@
+        "targets": {
+            "console": {
+                "type": "Console",
+                "layout": "${date}|${level:uppercase=true}|${message} ${exception:format=tostring}|${logger}|${all-event-properties}"
```
layout 裡頭寫的是什麼意思

cycumonk replied:
${data} 時間
${level:uppercase=true} 級別
${message} logger.Info(message);
${exception:format=tostring} 異常訊息 沒有就不會出現
${all-event-properties} 事件如果有添加props 例如username ip 可以在這裡印出來

miterfrants reviewed (commented):
checked

cycumonk-coder committed:
fix:
1.修改log檔存放的資料夾，改判斷若沒有資料夾在根目錄新增一個logs資料夾放log
2.修改LogManager關閉的時機，改由判斷app關閉在關閉Logmanager
